### PR TITLE
[SPARK-47069][PYTHON][CONNECT] Introduce `spark.profile.show/dump` for SparkSession-based profiling

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Introduce `spark.profile.show/dump` for SparkSession-based profiling for non-Spark-Connect. 

### Why are the changes needed?
SparkContext-based profiling has `sc.dump_profiles/show_profiles` for both perf and memory profiling.
Currently SparkSession-based has `spark.dump/showPerfProfiles` and `spark.dump/showMemoryProfiles` for perf and memory profiling separately.
It would be more consistent and user-friendly to consolidate them to a uniform interface as `spark.profile.dump/show`.

### Does this PR introduce _any_ user-facing change?
Yes. `spark.profile.show/dump` is supported, whereas (not-released yet) APIs below are removed
-`spark.dumpPerfProfiles`
-`spark.dumpMemoryProfiles`
-`spark.showPerfProfiles`
-`spark.showMemoryProfiles`

```py
>>> spark.conf.set("spark.sql.pyspark.udf.profiler", "perf")  # enable cProfiler
>>> 
>>> @udf("string")
... def f(x):
...       return str(x)
... 
>>> df = spark.range(10).select(f(col("id")))
>>> df.collect()
[Row(f(id)='0'), ...]
>>> spark.profile.show()
============================================================
Profile of UDF<id=2>
============================================================
...

>>> spark.profile.show(type="memory")
>>> spark.profile.show(type="perf")
============================================================
Profile of UDF<id=2>
============================================================
...

>>> spark.profile.show(2, type="perf")
============================================================
Profile of UDF<id=2>
============================================================
...

>>> spark.profile.show(2, type="memory")

```

### How was this patch tested?
Unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: python/pyspark/sql/profiler.py [hunk 1]
```diff
@@ -239,3 +241,72 @@ def _profile_results(self) -> "ProfileResults":
...
+
+        Parameters
+        ----------
+        type : str, optional
+            The profiler type, which can be either "perf" or "memory".
+        id : int, optional
+            A UDF ID to be shown. If not specified, all the results will be shown.
+        """
+        if type == "memory":
+            self.sparkSession.showMemoryProfiles(id)
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:270]: Shall we remove the old APIs? I think the new APIs are enough to have.

File: python/pyspark/sql/profiler.py [hunk 2]
```diff
@@ -239,3 +241,72 @@ def _profile_results(self) -> "ProfileResults":
...
+    """User-facing profile API. This instance can be accessed by
+    :attr:`spark.profile`.
+
+    .. versionadded: 4.0.0
+    """
+
+    def __init__(self, sparkSession: "SparkSession"):
+        self.sparkSession = sparkSession
+
+    def show(self, *, type: Optional[str] = None, id: Optional[int] = None) -> None:
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:256]: I prefer:
```suggestion
    def show(self, id: Optional[int] = None, *, type: Optional[str] = None) -> None:
```

```py
spark.profile.show()  # show all profile results
spark.profile.show(1)  # show the profile results for ID = 1
spark.profile.show(1, type="memory")  # show the memory profile results for ID = 1
spark.profile.show(type="memory")  # show all memory profile results
```

File: python/pyspark/sql/profiler.py [hunk 3]
```diff
@@ -239,3 +241,72 @@ def _profile_results(self) -> "ProfileResults":
...
+        else:
+            raise PySparkValueError(
+                error_class="VALUE_NOT_ALLOWED",
+                message_parameters={
+                    "arg_name": "type",
+                    "allowed_values": str(["perf", "memory"]),
+                },
+            )
+
+    def dump(self, path: str, *, type: Optional[str] = None, id: Optional[int] = None) -> None:
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:284]: ditto.
```suggestion
    def dump(self, path: str, id: Optional[int] = None, *, type: Optional[str] = None) -> None:
```

File: python/pyspark/sql/profiler.py [hunk 4]
```diff
@@ -239,3 +241,72 @@ def _profile_results(self) -> "ProfileResults":
...
+
+class Profile:
+    """User-facing profile API. This instance can be accessed by
+    :attr:`spark.profile`.
+
+    .. versionadded: 4.0.0
+    """
+
+    def __init__(self, sparkSession: "SparkSession"):
+        self.sparkSession = sparkSession
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:254]: It should take `ProfilerCollector`?

File: python/pyspark/sql/session.py [hunk 5]
```diff
@@ -906,6 +907,12 @@ def dataSource(self) -> "DataSourceRegistration":
 
         return DataSourceRegistration(self)
 
+    @property
+    def profile(self) -> "Profile":
```
Comment (ueshin) [inline python/pyspark/sql/session.py:911]: Need this for connect, too?

File: python/pyspark/sql/profiler.py [hunk 6]
```diff
@@ -239,3 +241,72 @@ def _profile_results(self) -> "ProfileResults":
         with self._lock:
             value = self._accumulator.value
             return value if value is not None else {}
+
+
+class Profile:
+    """User-facing profile API. This instance can be accessed by
```
Comment (HyukjinKwon) [inline python/pyspark/sql/profiler.py:247]: We should probably also fix the docs at `python/docs/source/reference`.

File: python/pyspark/sql/profiler.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/profiler.py:270]: 👍

Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py]: Sounds good!

File: python/pyspark/sql/profiler.py (hunk 2 above)
Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py]: Your suggestion makes more sense. Thanks!

File: python/pyspark/sql/profiler.py (hunk 3 above)
Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py]: Adjusted

File: python/pyspark/sql/profiler.py (hunk 4 above)
Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py]: Yes after removing the old APIs.

File: python/pyspark/sql/session.py (hunk 5 above)
Comment (xinrong-meng) [reply python/pyspark/sql/session.py:911]: I planned to make a follow-up PR for Connect. But now I think I'll just add Connect changes to this PR.

File: python/pyspark/sql/profiler.py (hunk 6 above)
Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py:247]: Good point! Would you mind if I file a follow-up pr for documentation? We might also want to edit the PySpark debugging guide together.

**Comment by xinrong-meng** [comment]
Adjusted PR description.

File: python/pyspark/sql/profiler.py [hunk 7]
```diff
@@ -239,3 +240,72 @@ def _profile_results(self) -> "ProfileResults":
...
+    :attr:`spark.profile`.
+
+    .. versionadded: 4.0.0
+    """
+
+    def __init__(self, profiler_collector: "ProfilerCollector"):
```
Comment (ueshin) [inline python/pyspark/sql/profiler.py:252]: nit: `"`s are necessary.

File: python/pyspark/sql/session.py [hunk 8]
```diff
@@ -76,6 +76,7 @@
     from pyspark.sql.udf import UDFRegistration
     from pyspark.sql.udtf import UDTFRegistration
     from pyspark.sql.datasource import DataSourceRegistration
+    from pyspark.sql.profiler import Profile
```
Comment (ueshin) [inline python/pyspark/sql/session.py:79]: Why is this here? Move to line 50?

File: python/pyspark/sql/session.py (hunk 5 above)
Comment (ueshin) [inline python/pyspark/sql/session.py:911]: I don't think `"`s are necessary.

File: python/pyspark/sql/connect/profiler.py [hunk 9]
```diff
@@ -39,3 +39,14 @@ def _profile_results(self) -> "ProfileResults":
     def _update(self, update: "ProfileResults") -> None:
         with self._lock:
             self._value = ProfileResultsParam.addInPlace(self._profile_results, update)
+
+
+class ConnectProfile(Profile):
```
Comment (ueshin) [inline python/pyspark/sql/connect/profiler.py:44]: I don't think this class is necessary. We can just use `Profile`?

File: python/pyspark/sql/session.py (hunk 8 above)
Comment (xinrong-meng) [reply python/pyspark/sql/session.py]: Adjusted. I noticed we intended to import inside a `property` as well as under TYPE_CHECKING, like `dataSource` and `udtf`, `udf`, etc. But now I believe that is not a convention.

File: python/pyspark/sql/profiler.py (hunk 7 above)
Comment (xinrong-meng) [reply python/pyspark/sql/profiler.py]: Adjusted.

File: python/pyspark/sql/session.py (hunk 5 above)
Comment (xinrong-meng) [reply python/pyspark/sql/session.py]: Adjusted.

**Review by HyukjinKwon — APPROVED** [review]

**Review by ueshin — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.

**Comment by xinrong-meng** [comment]
Thank you!
